datastream-server-api/src/main/java/com/linkedin/datastream/server/callbackstatus/CallbackStatusWithNonComparableOffsets.java

```diff
@@ -0,0 +1,84 @@
+/**
+ *  Copyright 2021 LinkedIn Corporation. All rights reserved.
+ *  Licensed under the BSD 2-Clause License. See the LICENSE file in the project root for license information.
+ *  See the NOTICE file in the project root for additional information regarding copyright ownership.
+ */
+package com.linkedin.datastream.server.callbackstatus;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Helper class to store the callback status of the inflight events with non comparable offsets.
+ *
+ * @param <T> Type of the non comparable checkpoint object internally used by the connector.
+ */
+public class CallbackStatusWithNonComparableOffsets<T> extends CallbackStatus<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CallbackStatusWithNonComparableOffsets.class);
+
+  // the last checkpoint-ed record's offset
+  protected T _currentCheckpoint = null;
+
+  // Hashset storing all the records which are yet to be acked
+  private final Set<T> _inFlight = Collections.synchronizedSet(new LinkedHashSet<>());
+
+  // Deque to store all the messages which are inflight until the last consumer checkpoint is made
+  private final Deque<T> _inFlightUntilLastConsumerCheckpoint = new ArrayDeque<>();
```

Comment:
One of the major drawback of using ArrayDeque is it keeps on doubling the capacity starting from 16 to 32,64,128,256... but it does not shrink it back.  This will result in two problems:

a. Large static memory allocations seen in the heap dump.
b. in case an ack is lost, this may keep on growing exponentially and can result in OOM.

Reply:
Ahh true, using LinkedList instead.